type_classes.md

```diff
@@ -251,41 +461,89 @@ instance : ToString Person where
+The second eval statement is using nested toString inference the inner one
+uses `ToString Person` which will give us `(String, String)` which is of type
+`String × String` so the outer one is using the predefined: `ToString String x String`.
+
 ## Numerals
 
-Numerals are polymorphic in Lean. You can use a numeral (e.g., `2`) to denote an element of any type that implements
-the type class `OfNat`.
+Numerals are polymorphic in Lean. You can use a numeral (e.g., `2`) to denote an element of any type that implements the type class `OfNat`.  Lean defines it this way:
+```lean
+# namespace Ex
+class OfNat (α : Type u) (n : Nat) where
+  ofNat : α
+# end Ex
+```
+
+Our previous type classes had only one Type parameter, notice this one has two parameters,
+the Type parameter `α` and an actual natural number `n`.  So we will create instances of this type
+class using syntax like this `instance : OfNat Float n` where n is some natural number.
+
+So now you can add an instance of `OfNat.ofNat` for your own custom `Rational` number type
+along with a `toString` implementation:
 ```lean
 structure Rational where
   num : Int
   den : Nat
-  inv : den ≠ 0
 
 instance : OfNat Rational n where
-  ofNat := { num := n, den := 1, inv := by decide }
+  ofNat := { num := n, den := 1 }
+
+#check (2 : Rational) -- Rational
 
 instance : ToString Rational where
   toString r := s!"{r.num}/{r.den}"
 
 #eval (2 : Rational) -- 2/1
-
-#check (2 : Rational) -- Rational
-#check (2 : Nat)      -- Nat
+#check (2 : Nat)     -- Nat
+#check 2             -- Still defaults to Nat
 ```
-Lean elaborates the terms `(2 : Nat)` and `(2 : Rational)` as
-`OfNat.ofNat Nat 2 (instOfNatNat 2)` and
-`OfNat.ofNat Rational 2 (instOfNatRational 2)` respectively.
+Lean elaborates the terms `(2 : Nat)` as `OfNat.ofNat Nat 2 (instOfNatNat 2)`
+where the `instOfNatNat` function is automatically defined by the system
+in the instance of `OfNat Nat`.
+
+Similarly `(2 : Rational)` is elaborated as
+`OfNat.ofNat Rational 2 (instOfNatRational 2)`.
 We say the numerals `2` occurring in the elaborated terms are *raw* natural numbers.
-You can input the raw natural number `2` using the macro `nat_lit 2`.
+You can input the raw natural number `2` using the macro `nat_lit 2`:
 ```lean
 #check nat_lit 2  -- Nat
 ```
-Raw natural numbers are *not* polymorphic.
+Raw natural numbers are *not* polymorphic, only unelaborated numerals are polymorphic.
+So `#check ((2 : Nat) : Rational)` fails with a type mismatch.
+
+The `OfNat` instance is parametric on the numeral. So, you can define
+instances for particular numerals. The second argument is often a
+variable as in the example above where we use the variable `n` in
+`OfNat Rational n`, or a *raw* natural number like this:
+
+```lean
+# structure Rational where
+#   num : Int
+#   den : Nat
+
+instance : OfNat Rational 2 where
+  ofNat := { num := 3, den := 4 }
+```
+
+BUGBUG: but what does this mean and how do you use it?
```

Comment:
It means that you can explain two things to lean:
- How to turn specific number literals into a value of a certain type, for example the instance here `OfNat Rational 2` explains to lean that if the number literal `2` shows up when a `Rational` is required it can use the value `{ num := 3, den := 4 }` in this place
- How to turn *any* numbe literal into a value of a certain type, this would be achieved by writing an instance of type `OfNat Rational n`, for example:

```lean
instance : OfNat Rational n where
  ofNat := { num := n, den := 1}
```
would allow lean to turn any number literal `n` into an appropriate `Rational` value.